### DISCUSSION TITLE
ValidationInfo, field order, and missing values

### DISCUSSION THREAD

**Discussion by amacfie-tc** [pydantic/pydantic#8309] [Question]
Consider this code
```python
from pydantic import (
    BaseModel,
    ValidationError,
    ValidationInfo,
    field_validator,
)


class UserModel(BaseModel):
    id: int
    name: str

    @field_validator("name")
    @classmethod
    def name_must_contain_space(cls, v: str, info: ValidationInfo) -> str:
        print(info.data["id"])
        if " " not in v:
            raise ValueError("must contain a space")
        return v.title()


print(UserModel(id=1, name="John Doe"))  # success
print(UserModel(name="John Doe"))  # KeyError: 'id' instead of "Field required"
```

The [docs](https://docs.pydantic.dev/latest/concepts/validators/#field-validators) say

> If you want to access values from another field inside a @field_validator, this may be possible using ValidationInfo.data, which is a dict of field name to field value. Validation is done in the order fields are defined, so you have to be careful when using ValidationInfo.data to not access a field that has not yet been validated/populated — in the code above, for example, you would not be able to access info.data['id'] from within name_must_contain_space.

If validation is done in the order the fields are defined, shouldn't `id` be validated first and fail if it's missing? How can we reference other fields from a field validator with a guarantee that they've already been validated?

**Comment by Viicos** [comment]
I think this is fixed by https://github.com/pydantic/pydantic/pull/8191

**Comment by amacfie-tc** [comment]
The answer appears to be in the [docs for v1](https://docs.pydantic.dev/1.10/usage/validators/#:~:text=If%20validation%20fails%20on%20another%20field):

>If validation fails on another field (or that field is missing) it will not be included in `values`

(where `values` is the v1 equivalent of `info.data`)
